### PR TITLE
Add IterableOps#tapEach method

### LINKED ISSUE #11098: Add method View#peek
`View`s could have a method `peek` which would allow the injection of a side-effecting function (or indeed, breakpoint) in the middle of the view's pipeline. Inspired by [Java 8 Stream#peek method](https://docs.oracle.com/javase/8/docs/api/?java/util/stream/Stream.html). The signature would be something like:

```scala
def peek[U](f: A => U): View[A] = new View.Peek(this, f)
```

and its usage would be like

```scala
     View("one", "two", "three", "four")
         .filter(_.length > 3)
         .peek(e => println("Filtered value: " + e))
         .map(_.toUpperCase)
         .peek(e => println("Mapped value: " + e))
         .to(Seq)
```

### PR DESCRIPTION
**Pull request by joshlemer**
This closes scala/bug#11098

This adds `tapEach` method to `IterableOps` which allows you to write:

```scala
View(1,2,3)
  .tapEach(println)
  .tapEach(println)
  .to(Seq)
// prints: 
1
1
2
2
3
3
```

Or 
```scala
List(1,2,3)
  .tapEach(println)
  .tapEach(println)
// prints: 
1
2
3
1
2
3
```

### CODE REVIEW

File: src/library/scala/collection/Iterable.scala
Comment (NthPortal) [inline src/library/scala/collection/Iterable.scala]: you could potentially optimise this for strict collections by doing

```scala
  def tapEach[U](f: A => U): C = knownSize match {
    case 0          => this
    case n if n < 0 => fromSpecific(new View.TapEach(this, f))
    case n if n > 0 =>
      foreach(f)
      this
```

Comment (NthPortal) [inline src/library/scala/collection/Iterable.scala]: I'm also not sure if we really need a `View.TapEach`, or can just do `map(a => { f(a); a })` instead of `fromSpecific(new View.TapEach(this, f))`

Comment (NthPortal) [reply src/library/scala/collection/Iterable.scala]: it's not perfect (e.g. it misses `List`), but it might be better than nothing?

Comment (joshlemer) [reply src/library/scala/collection/Iterable.scala]: You're right, I'll change this

Comment (joshlemer) [reply src/library/scala/collection/Iterable.scala]: I think that this size optimization could result in surprising behaviour if the underlying collection is mutated but lazy. In that case, when the empty collection has elements inserted, and then the lazy collection forces, `f` will not be called. I do not know if there's actually any collections like this however...

**Comment by joshlemer** [comment]
@NthPortal ready for re-review

**Comment by javax-swing** [comment]
I think this will need to be specialised for LazyList, since LazyList overrides `map`, I don’t think it plays nicely with views

**Comment by joshlemer** [comment]
@javax-swing Added some tests [here](https://github.com/scala/scala/pull/7124/files#diff-c8e80d186a09bf13da87d13cb4913e54R376) for LazyList#tapEach

**Comment by NthPortal** [comment]
The fact that you need a specialization for `LazyList` is why I think it would be better to implement this as a call to `map`, rather than as `fromSpecific(View.Map(...))`; once you have `map` doing the right thing, `tapEach` will be correct for free

**Comment by NthPortal** [comment]
I'm wondering if it would be a good idea to have a `StrictTapEach` trait to mix into strict collections; e.g.

```scala
trait StrictTapEach[A] {
  self: IterableOps[A, AnyConstr, _] =>
  override def tapEach[U](f: A => U): this.type = {
    foreach(f)
    this
  }
}
```

(if `this.type` isn't valid for for some reason, you could add a `C` type parameter to the self type)

**Comment by joshlemer** [comment]
>I'm wondering if it would be a good idea to have a StrictTapEach trait to mix into strict collections; 
@NthPortal   

Actually I just put the implementation in `StrictOptimizedIterableOps` which all the strict iterables extend

File: src/library/scala/collection/StrictOptimizedIterableOps.scala
Comment (NthPortal) [inline src/library/scala/collection/StrictOptimizedIterableOps.scala:255]: can't return `this`?

Comment (joshlemer) [reply src/library/scala/collection/StrictOptimizedIterableOps.scala:255]: Unfortunately, `StrictOptimizedSeqOps[..., C]` does not conform to `C`. I could play around with introducing a self-type to it, `self: C => ` and see if there are any issues.

Comment (joshlemer) [reply src/library/scala/collection/StrictOptimizedIterableOps.scala:255]: Ok, It seems that this solution does solve it, as long as you also enforce the `self: C =>` type on all extending `StrictOptimized{Seq,Set,Map,SortedMap, SortedSet}Ops`. I'd like to get @julienrf 's opinion though if he forsees any issues with that constraint, maybe it's intended to be extendable by non-`C` things too

File: src/library/scala/collection/StrictOptimizedIterableOps.scala
Comment (julienrf) [inline src/library/scala/collection/StrictOptimizedIterableOps.scala]: I think it would be simpler to return `coll` here (and keep using `C` as the return type).

File: test/junit/scala/collection/immutable/LazyListTest.scala
Comment (NthPortal) [inline test/junit/scala/collection/immutable/LazyListTest.scala]: you can make the `LazyList` independently lazy by doing `LazyList.tabulate(5)(_ + 1)`

File: test/junit/scala/collection/immutable/LazyListTest.scala
Comment (NthPortal) [inline test/junit/scala/collection/immutable/LazyListTest.scala:1013]: Ideally there should be a test `def tapEach_properlyLazy(): Unit` right after `def map_properlyLazy(): Unit` which asserts laziness; it can be something like the following:

```scala
   @Test
   def tapEach_properlyLazy(): Unit = {
     val op = lazyListOp(_.tapEach(_ + 1)) // or some other invocation of `tapEach` here
     assertRepeatedlyFullyLazy(op)
     assertLazyTailWhenHeadEvaluated(op)
     assertLazyHeadWhenTailEvaluated(op)
     assertLazyHeadWhenNextHeadEvaluated(op)
   }
```

You can also have this method to check that the behaviour is correct, but there are a lot of `assert<Something>` methods you may find helpful for checking laziness (you can look at the other `<op>_properlyLazy()` methods for examples).

**Comment by joshlemer** [comment]
@NthPortal your suggestions are implemented now

File: src/library/scala/collection/Iterator.scala
Comment (NthPortal) [inline src/library/scala/collection/Iterator.scala]: we probably don't need three copies of this scaladoc

File: test/junit/scala/collection/IteratorTest.scala
Comment (NthPortal) [inline test/junit/scala/collection/IteratorTest.scala]: should these use `locally`?

**Review by NthPortal — APPROVED** [review]
Nice!

**Comment by NthPortal** [comment]
IMO, this is ready to merge (once the conflict is resolved). @julienrf do you agree?

**Comment by joshlemer** [comment]
I think an argument could be made to manually override each strict
collection with an implementation that returns `this.type`, what do you
think?

**Comment by NthPortal** [comment]
@joshlemer I'm not sure if I'm behind increasing the maintenance burden by adding a whole bunch of overrides that just call super. If there's a way to do it generically so that you only need at most a few overrides, that sounds good to me, but I'm not sure it's worth it to do it for every strict collection.

File: src/library/scala/collection/StrictOptimizedIterableOps.scala
Comment (szeiger) [reply src/library/scala/collection/StrictOptimizedIterableOps.scala:255]: The exceptions used to be `StringOps` and `ArrayOps`. All regular collection types extend their `C` parameter.  Can we make any further simplifications if we always enforce this? I don't think it's worth making this change specifically for `tapEach`. There are plenty of other methods that return `C`. In most cases you abstract over consuming a collection where this is not an issue. If you also need to produce a more specific collection type you need to use the proper abstraction anyway (unless we can change it in a more generic way).

File: src/library/scala/collection/IterableOnce.scala
Comment (szeiger) [inline src/library/scala/collection/IterableOnce.scala]: Is there a non-typical use? I would omit "typically".

Comment (joshlemer) [reply src/library/scala/collection/IterableOnce.scala]: 👍

File: src/library/scala/collection/StrictOptimizedIterableOps.scala
Comment (joshlemer) [reply src/library/scala/collection/StrictOptimizedIterableOps.scala:255]: @szeiger I can't think of any other improvements or simplifications that could be made if it had `{ self: C => ` unfortunately.

File: src/library/scala/collection/immutable/LazyList.scala
Comment (SethTisue) [inline src/library/scala/collection/immutable/LazyList.scala:383]: should this be `mapConserve` instead of `map`?

Comment (joshlemer) [reply src/library/scala/collection/immutable/LazyList.scala:383]: I've never heard of `mapConserve` and it doesn't seem to be available on `LazyList`, what's that?

Comment (SethTisue) [reply src/library/scala/collection/immutable/LazyList.scala:383]: it avoids consing up a new list if all the items are unchanged

but ohhhh I see `mapConserve` only exists on `List`. maybe `LazyList` could or should have it, nbd, it's pretty obscure stuff, though. it's used a fair amount within the compiler.

nm I guess

Comment (NthPortal) [reply src/library/scala/collection/immutable/LazyList.scala:383]: I don't think you can have it on `LazyList` - in order to tell if you can return `this`, you need to evaluate the whole mapped list, which violates the laziness

**Comment by adriaanm** [comment]
@szeiger do I read your review correctly as "not convinced yet?"

**Comment by NthPortal** [comment]
Is there a reason for this not to be merged?

**Comment by joshlemer** [comment]
@nthportal I don't believe so

**Comment by adriaanm** [comment]
ping @szeiger

**Review by SethTisue — APPROVED** [review]

File: src/library/scala/collection/Iterator.scala
Comment (szeiger) [inline src/library/scala/collection/Iterator.scala:866]: This should override `knownSize`

**Review by szeiger — COMMENTED** [review]
Other than the missing `knownSize`, this looks good.

File: src/library/scala/collection/Iterable.scala
Comment (NthPortal) [inline src/library/scala/collection/Iterable.scala:822]: Could I get others' opinions on whether or not this should be implemented as a call to `map` instead of `fromSpecific(new View.Map)`? I think that the former simplifies/reduces the work for those implementing collections which need to override `map`, but I am curious to hear others' thoughts.

I don't think the decision should hold up the PR - it's an easy change to make in its own PR if needed.

Comment (julienrf) [reply src/library/scala/collection/Iterable.scala:822]: I think using `map` is better, indeed.

**Comment by lrytz** [comment]
ping @joshlemer

File: src/library/scala/collection/Iterable.scala
Comment (joshlemer) [reply src/library/scala/collection/Iterable.scala:822]: Unfortunately, calling `map { a => f(a); a }` will only return a `CC[A]`, rather than the more specific `C`.

File: src/library/scala/collection/Iterator.scala
Comment (joshlemer) [reply src/library/scala/collection/Iterator.scala:866]: Done!

**Comment by szeiger** [comment]
Strange compiler failure in the Travis build: https://travis-ci.org/scala/scala/jobs/449897364. I've triggered a new build.

**Comment by lrytz** [comment]
Travis should be ignored for the time being, https://github.com/scala/scala-dev/issues/570

File: src/library/scala/collection/Iterable.scala
Comment (NthPortal) [reply src/library/scala/collection/Iterable.scala:822]: That is interesting and unfortunate. I'm torn on whether or not it would still be better to use `map` and cast the result. I'm leaning towards not.

**Comment by szeiger** [comment]
I see. In this case it just needs to be squashed because the first commit on its own failed.

File: src/library/scala/collection/Iterable.scala
Comment (SethTisue) [reply src/library/scala/collection/Iterable.scala:822]: I asked @szeiger to look at this and explain it to me.

Josh's point is that we can't typecast our way out of this. `CC` and `C[A]` can be different collection types entirely:

```
scala 2.13.0-M5> (Map(1 -> "foo"): Iterable[(Int, String)]).map(x => x)
res5: Iterable[(Int, String)] = List((1,foo))

scala 2.13.0-M5> (collection.immutable.BitSet(0, 5, 9): Iterable[Int]).map(_ + 1)
res0: Iterable[Int] = Set(1, 6, 10)
```

(in an earlier edit, I'd included a `String` example where I "upcast" it to `Iterable[Char]`, but that's not really the same, because it's an implicit conversion, not an upcast)

Comment (SethTisue) [reply src/library/scala/collection/Iterable.scala:822]: @szeiger asks whether using `filter` instead was considered. the problem with `map` is that `map` loses the information that the element type is unchanged, but with `filter` we have that

```diff
-  override def tapEach[U](f: A => U): C = fromSpecific(new View.Map(this, { a: A => f(a); a }))
+  override def tapEach[U](f: A => U): C = filter{a: A => f(a); true}
```

neither Stefan and I can think of a downside. there's no significant upside, either, that we can see, it just seems a touch cleaner. thoughts? (the `filter` version compiles, but I ran out of time to try running the tests on it)

Comment (joshlemer) [reply src/library/scala/collection/Iterable.scala:822]: @SethTisue Wow!  Very sneaky! Some downsides (which may have workarounds):

* filter can (as far as the compiler is concerned) change the length of the `Iterable`. So the consequences of this are that, perhaps non-strict Seq's like SeqView and LazyList will have to then go through each element in order to fulfill a call to `apply(100)`, whereas before the `filter` call, they may have been able to independently evaluate/access element `100` immediately. 
* The above point means that `IndexedSeqView#filter` cannot return an `IndexedSeqView`

Comment (NthPortal) [reply src/library/scala/collection/Iterable.scala:822]: To add on one more point, `filter` also destroys independent laziness of elements (currently only a thing for `LazyList`, but could potentially be a thing for some lazy `IndexedSeq` in the future).

Comment (SethTisue) [reply src/library/scala/collection/Iterable.scala:822]: @joshlemer okay, sounds like it’s better as-is then

@NthPortal I don’t follow?

Comment (NthPortal) [reply src/library/scala/collection/Iterable.scala:822]: @SethTisue 

```scala
scala> def list = LazyList.tabulate(10)(i => i)
list: scala.collection.immutable.LazyList[Int]

scala> def f(i: Int) = println(i)
f: (i: Int)Unit

scala> list.map(i => {f(i); i})(5)
5
res0: Int = 5

scala> list.filter(i => {f(i); true})(5)
0
1
2
3
4
5
res1: Int = 5
```

Comment (NthPortal) [reply src/library/scala/collection/Iterable.scala:822]: I retract all complaints, and I'm happy with `fromSpecific(new View.Map(...))`

**Comment by SethTisue** [comment]
@joshlemer rebase and squash?

**Comment by joshlemer** [comment]
Done! @SethTisue

**Comment by SethTisue** [comment]
gah, sorry Josh, needs one more rebase.

I tried to use this the other day and was surprised to find it hadn't been merged yet :-)

**Comment by joshlemer** [comment]
🤞 @SethTisue

**Pull request merged**
